private/pkg/manifest/module.go

```diff
@@ -0,0 +1,321 @@
+	var config blobSetOptions
+	for _, option := range opts {
+		option(&config)
+	}
+	digestToBlobs := make(map[string]Blob, len(blobs))
+	for _, b := range blobs {
+		digestStr := b.Digest().String()
+		if config.validateContent {
+			existingBlob, alreadyPresent := digestToBlobs[digestStr]
+			if alreadyPresent {
+				equalContent, err := BlobEqual(ctx, b, existingBlob)
+				if err != nil {
+					return nil, fmt.Errorf("compare duplicated blobs with digest %q: %w", digestStr, err)
+				}
+				if !equalContent {
+					return nil, fmt.Errorf("duplicated blobs with digest %q have different contents", digestStr)
+				}
+			}
+		}
+		digestToBlobs[digestStr] = b
+	}
+	return &BlobSet{digestToBlob: digestToBlobs}, nil
+}
+
+// BlobFor returns the blob for the passed digest string, or nil, ok=false if
+// the digest has no blob in the set.
+func (s *BlobSet) BlobFor(digest string) (Blob, bool) {
+	blob, ok := s.digestToBlob[digest]
+	if !ok {
+		return nil, false
+	}
+	return blob, true
+}
+
+// Blobs returns the blobs in the set as an array.
+func (s *BlobSet) Blobs() []Blob {
+	blobs := make([]Blob, 0, len(s.digestToBlob))
+	for _, b := range s.digestToBlob {
+		blobs = append(blobs, b)
+	}
+	return blobs
+}
+
+// NewDigestFromBlobHash maps a module Hash to a digest.
+func NewDigestFromBlobHash(hash *modulev1alpha1.Hash) (*Digest, error) {
+	if hash == nil {
+		return nil, fmt.Errorf("nil hash")
+	}
+	dType, ok := hashKindToDigestType[hash.Kind]
+	if !ok {
+		return nil, fmt.Errorf("unsupported hash kind: %s", hash.Kind.String())
+	}
+	return NewDigestFromBytes(dType, hash.Digest)
+}
+
+// NewMemoryBlobFromReader creates a memory blob from content, which is read
+// until completion. The returned blob contains all bytes read. If you are using
+// this in a loop, you might better use NewMemoryBlobFromReaderWithDigester so
+// you can reuse your digester.
+func NewMemoryBlobFromReader(content io.Reader) (Blob, error) {
+	digester, err := NewDigester(DigestTypeShake256)
+	if err != nil {
+		return nil, err
+	}
+	return NewMemoryBlobFromReaderWithDigester(content, digester)
+}
+
+// NewMemoryBlobFromReaderWithDigester creates a memory blob from content with
+// the passed digester. The content is read until completion. The returned blob
+// contains all bytes read.
+func NewMemoryBlobFromReaderWithDigester(content io.Reader, digester Digester) (Blob, error) {
+	var contentInMemory bytes.Buffer
+	tee := io.TeeReader(content, &contentInMemory)
+	digest, err := digester.Digest(tee)
+	if err != nil {
+		return nil, err
+	}
+	return &memoryBlob{
+		digest:  *digest,
+		content: contentInMemory.Bytes(),
+	}, nil
+}
+
+// BlobEqual returns true if blob a is the same as blob b. The digest is
+// checked for equality and the content bytes compared.
+//
+// An error is returned if an unexpected I/O error occurred when opening,
+// reading, or closing either blob.
+func BlobEqual(ctx context.Context, a, b Blob) (_ bool, retErr error) {
+	const blockSize = 4098
```

Comment:
Any reason we're using this block size? Should it be documented?

Reply:
It's a number. Is it a reasonable one? I don't know. It's one disk block size. It's less than one Ethernet frame. 🤷 

IMHO I wouldn't document it in the function docs; it's not behavior that should be relied upon. Maybe as a note on this line that it's open for tweaking?